File: old/src/jsMain/kotlin/org/baaahs/views/index/components/Hero.kt
package org.baaahs.views.index.components

import csstype.AlignItems
import csstype.Auto
import csstype.BackgroundColor
import csstype.BackgroundImage
import csstype.BackgroundRepeat
import csstype.ClassName
import csstype.Color
import csstype.Display
import csstype.FlexDirection
import csstype.FontFamily
import csstype.FontStyle
import csstype.None
import csstype.Position
import csstype.TextDecoration
import csstype.deg
import csstype.dropShadow
import csstype.em
import csstype.integer
import csstype.pct
import csstype.px
import csstype.rem
import csstype.rotate
import csstype.translate3d
import emotion.react.css
import externals.mui.material.styles.alpha
import js.core.ArrayLike
import js.core.asList
import js.core.jso
import mui.material.Box
import mui.material.Button
import mui.material.ButtonColor
import mui.material.ButtonVariant
import mui.material.Container
import mui.material.PaletteMode
import mui.material.Size
import mui.material.Typography
import mui.material.styles.Theme
import mui.material.styles.TypographyVariant
import mui.material.styles.useTheme
import mui.material.useMediaQuery
import mui.system.Breakpoint
import mui.system.sx
import org.baaahs.util.breakpoints
import org.baaahs.util.marginX
import org.baaahs.util.paddingY
import org.baaahs.util.sp
import org.baaahs.util.useComponent
import react.FC
import react.Props
import react.PropsWithChildren
import react.dom.html.ReactHTML.a
import react.dom.html.ReactHTML.img
import react.dom.html.ReactHTML.p
import react.dom.html.ReactHTML.span
import react.dom.svg.ReactSVG.path
import react.dom.svg.ReactSVG.svg
import web.dom.document
import web.html.Loading

data class Image(val cover: String, val coverDark: String)

private external interface HoverableLetter : PropsWithChildren {
    var letterSelector: String
}

private val hoverableLetter = FC<HoverableLetter> { props ->
    span {
        onMouseOver = {
            document.querySelectorAll(props.letterSelector).forEach {
                it.classList.add("baaahs-letter-active")
            }
            console.log("Select " + props.letterSelector)
        }
        onMouseOut = {
            document.querySelectorAll(props.letterSelector).forEach {
                it.classList.remove("baaahs-letter-active")
            }
            console.log("Deselect " + props.letterSelector)
        }
        +props.children
    }
}

val imageGroups = listOf(
    listOf(
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img1.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img1--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img4.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img4--dark.png"
        )
    ),
    listOf(
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img13.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img13--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img10.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img10--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img7.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img7--dark.png"
        )
    ),
    listOf(
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img6.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img6--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img24.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img24--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img17.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img17--dark.png"
        ),
        Image(
            "https://assets.maccarianagency.com/screenshots/the-front/img12.png",
            "https://assets.maccarianagency.com/screenshots/the-front/img12--dark.png"
        )
    )
)

val Hero = FC<Props> {
    val theme = useTheme<Theme>()
    val isMd = useMediaQuery(theme.breakpoints.up(Breakpoint.md), jso {
        defaultMatches = true
    })

    Box {
        sx {
            backgroundImage =
                "linear-gradient(to bottom, ${alpha(theme.palette.background.paper, 0)}, ${
                    alpha(
                        theme.palette.asDynamic().alternate.main,
                        1
                    )
                } 100%)" as BackgroundImage
            backgroundRepeat = BackgroundRepeat.repeatX
            position = Position.relative

        }

        Box {
            sx {
                paddingY = breakpoints {
                    xs = 0.sp
                    sm = 4.rem
                    md = 8.rem
                }
            }

            Container {
                id = "hero"

                Box {
                    sx { maxWidth = breakpoints { xs = 1.sp; sm = 50.pct } }

                    Typography {
                        variant = TypographyVariant.h2
                        gutterBottom = true
                        sx {
                            fontSize = breakpoints {
                                xs = 4.5.rem; sm = 4.5.rem; md = 4.5.rem; lg = 5.75.rem; xl = 5.75.rem
                            }
                            color = Color("text.primary")
                            fontFamily = "Smooth Circulars" as FontFamily
                        }

                        hoverableLetter { letterSelector = ".baaahs-letter-b"; +"B" }
                        hoverableLetter { letterSelector = ".baaahs-letter-a1"; +"A" }
                        hoverableLetter { letterSelector = ".baaahs-letter-a2"; +"A" }
                        hoverableLetter { letterSelector = ".baaahs-letter-a3"; +"A" }
                        hoverableLetter { letterSelector = ".baaahs-letter-h"; +"H" }
                        hoverableLetter { letterSelector = ".baaahs-letter-s"; +"S" }
                    }

                    Typography {
                        variant = TypographyVariant.h3
                        gutterBottom = true
                        sx {
                            fontSize = breakpoints {
                                xs = 0.85.rem; sm = 0.85.rem; md = 0.85.rem; lg = 0.85.rem; xl = 0.85.rem
                            }
                            color = Color("text.primary")
                            fontFamily = "Smooth Circulars" as FontFamily
                            marginTop = (-20).px

                            "& .baaahs-letter-active" {
                                textDecoration = TextDecoration.underline
                                filter = dropShadow(0.px, 0.px, 1.em)
                            }
                        }

                        span { css(ClassName("baaahs-letter-b")) {}; +"Big" }
                        +"-"
                        span { css(ClassName("baaahs-letter-a1")) {}; +"Ass" }
                        +" "
                        span { css(ClassName("baaahs-letter-a2")) {}; +"Amazingly" }
                        +" "
                        span { css(ClassName("baaahs-letter-a3")) {}; +"Awesome" }
                        +" "
                        span { css(ClassName("baaahs-letter-h")) {}; +"Homosexual" }
                        +" "
                        span { css(ClassName("baaahs-letter-s")) {}; +"Sheep" }
                    }

                    Typography {
                        variant = TypographyVariant.h6
                        useComponent(p)
                        sx {
                            color = Color("text.secondary")
                            fontWeight = integer(400)

                            "& .baaahs-letter-active" {
                                fontStyle = FontStyle.italic
                                textDecoration = TextDecoration.underline
                            }
                        }

                        +"BAAAHS (the "
                        span { css(ClassName("baaahs-letter-b")) {}; +"Big" }
                        +"-"
                        span { css(ClassName("baaahs-letter-a1")) {}; +"Ass" }
                        +" "
                        span { css(ClassName("baaahs-letter-a2")) {}; +"Amazingly" }
                        +" "
                        span { css(ClassName("baaahs-letter-a3")) {}; +"Awesome" }
                        +" "
                        span { css(ClassName("baaahs-letter-h")) {}; +"Homosexual" }
                        +" "
                        span { css(ClassName("baaahs-letter-s")) {}; +"Sheep" }
                        +") is a mutant vehicle, a mobile disco, and a penetrable social statement."
                    }

                    Box {
                        sx {
                            display = Display.flex
                            flexDirection = breakpoints {
                                xs = FlexDirection.column
                                sm = FlexDirection.row
                            }
                            alignItems = breakpoints {
                                xs = AlignItems.stretch
                                sm = AlignItems.flexStart
                            }
                            marginTop = 4.sp
                        }

                        Button {
                            useComponent(a)
                            variant = ButtonVariant.contained
                            color = ButtonColor.primary
                            size = Size.large
                            fullWidth = if (isMd) false else true
                            href = "/crew"
                            +"Join the flock"
                        }
                        Box {
                            sx {
                                marginTop = breakpoints { xs = 2.sp; sm = 0.sp }
                                marginLeft = breakpoints { sm = 2.sp }
                                width = breakpoints { xs = 100.pct; md = Auto.auto }
                            }

                            Button {
                                useComponent(a)
                                href = "/about"
                                variant = ButtonVariant.outlined
                                color = ButtonColor.primary
                                size = Size.large
                                fullWidth = !isMd
                                +"Learn more"
                            }
                        }
                    }
                }
            }
            Box {
                sx {
                    transform = rotate((-20).deg)
                    display = breakpoints { xs = None.none; sm = Display.block }
                }

                Box {
                    sx {
                        display = Display.flex
                        width = 50.rem
                        left = 50.pct
                        top = 0.px
                        position = Position.absolute
                        transform = translate3d(20.pct, (-50).pct, 0.pct)

                    }

                    imageGroups.forEachIndexed { i, images ->
                        Box {
                            key = i.toString()
                            sx {
                                marginTop = breakpoints { sm = (-i * 16).sp }
                                marginX = 1.sp
                            }

                            images.forEachIndexed { j, image ->
                                Box {
                                    key = j.toString()
                                    sx {
                                        padding = 1.sp
                                        backgroundColor = "background.paper" as BackgroundColor
                                        borderRadius = 2.sp
                                        boxShadow = 3.sp
                                        marginTop = 2.sp
                                    }

                                    Box {
                                        useComponent(img) {
                                            loading = Loading.lazy
                                            src =
                                                if (theme.palette.mode == PaletteMode.dark) image.coverDark else image.cover
                                        }
                                        sx {
                                            height = 1.sp
                                            width = 1.sp
                                            maxWidth = 320.px
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        Box {
            useComponent(svg) {
                preserveAspectRatio = "none"
                xmlns = "http://www.w3.org/2000/svg"
                x = 0.0
                y = 0.0
                viewBox = "0 0 1920 100.1"
            }
            sx {
                width = 100.pct

                marginBottom = theme.spacing(-1)

            }

            path {
                fill = theme.palette.background.paper
                d = "M0,0c0,0,934.4,93.4,1920,0v100.1H0L0,0z"
            }
        }
    }
}

fun <T> ArrayLike<T>.forEach(block: (T) -> Unit) = asList().forEach { block(it) }

File: src/views/IndexView/components/Hero/Hero.js
import React from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
// import Button from '@mui/material/Button';
// import useMediaQuery from '@mui/material/useMediaQuery';
import { alpha, useTheme } from '@mui/material/styles';

import Container from 'components/Container';

const images = [
  {
    group: [
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/2013_sheep.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/2015_sheep.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/shady_xian_discoballs.jpeg',
        coverDark: undefined,
      },
    ],
  },
  {
    group: [
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/2019_sheep.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/rainbow_baaahs.png',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/sunset_balls_sheep.jpeg',
        coverDark: undefined,
      },
    ],
  },
  {
    group: [
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/2019_sheep_2.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/build_sheep.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://storage.googleapis.com/static.baaahs.org/kinky_sheep.jpeg',
        coverDark: undefined,
      },
      {
        cover:
          'https://assets.maccarianagency.com/screenshots/the-front/img12.png',
        coverDark: undefined,
      },
    ],
  },
];

const HoverableLetter = ({ letterSelector, children }) => {
  return (
    <span
      onMouseOver={() => {
        document.querySelectorAll(letterSelector).forEach((el) => {
          el.classList.add('baaahs-letter-hover');
        });
      }}
      onMouseOut={() => {
        document.querySelectorAll(letterSelector).forEach((el) => {
          el.classList.remove('baaahs-letter-hover');
        });
      }}
    >
      {children}
    </span>
  );
};

const Hero = () => {
  const theme = useTheme();
  // const isMd = useMediaQuery(theme.breakpoints.up('md'), {
  //   defaultMatches: true,
  // });

  return (
    <Box
      sx={{
        backgroundImage: `linear-gradient(to bottom, ${alpha(
          theme.palette.background.paper,
          0,
        )}, ${alpha(theme.palette.alternate.main, 1)} 100%)`,
        backgroundRepeat: 'repeat-x',
        position: 'relative',
      }}
    >
      <Box paddingY={{ xs: 0, sm: '4rem', md: '8rem' }}>
        <Container>
          <Box maxWidth={{ xs: 1, sm: '50%' }}>
            <Typography
              component="h2"
              gutterBottom={true}
              width={{ xs: 300, md: 580 }}
              marginBottom={{ xs: 1, sm: 2 }}
              color="text.primary"
              fontFamily={'Smooth Circulars'}
              fontSize={{
                xs: '3.7rem',
                sm: '4rem',
                md: '4.5rem',
                lg: '5.75rem',
                xl: '5.75.rem',
              }}
              sx={{ cursor: 'default' }}
            >
              <HoverableLetter letterSelector=".baaahs-letter-b">
                B
              </HoverableLetter>
              <HoverableLetter letterSelector=".baaahs-letter-a1">
                A
              </HoverableLetter>
              <HoverableLetter letterSelector=".baaahs-letter-a2">
                A
              </HoverableLetter>
              <HoverableLetter letterSelector=".baaahs-letter-a3">
                A
              </HoverableLetter>
              <HoverableLetter letterSelector=".baaahs-letter-h">
                H
              </HoverableLetter>
              <HoverableLetter letterSelector=".baaahs-letter-s">
                S
              </HoverableLetter>
            </Typography>

            <Typography
              variant="h3"
              marginTop="-20px"
              marginBottom="2em"
              gutterBottom={true}
              component="p"
              color="text.primary"
              fontFamily={'Smooth Circulars'}
              fontSize={{
                xs: '.55rem',
                sm: '.59rem',
                md: '.67rem',
                lg: '.85rem',
                xl: '.85rem',
              }}
              whiteSpace="nowrap"
              sx={{
                fontWeight: 400,
                cursor: 'default',
                '& .baaahs-letter-hover': {
                  fontWeight: 700,
                  fontStyle: 'italic',
                },
              }}
            >
              <span className="baaahs-letter-b">Big</span>-
              <span className="baaahs-letter-a1">Ass</span>
              <span> </span>
              <span className="baaahs-letter-a2">Amazingly</span>
              <span> </span>
              <span className="baaahs-letter-a3">Awesome</span>
              <span> </span>
              <span className="baaahs-letter-h">Homosexual</span>
              <span> </span>
              <span className="baaahs-letter-s">Sheep</span>
              <span> </span>
            </Typography>

            <Typography variant="h6" component="p" color="text.primary">
              The Big-Ass Amazingly Awesome Homosexual Sheep (BAAAHS) is a
              mutant vehicle, a mobile disco, and a penetrable social statement.
            </Typography>
            {/*
            <Box
              display="flex"
              flexDirection={{ xs: 'column', sm: 'row' }}
              alignItems={{ xs: 'stretched', sm: 'flex-start' }}
              marginTop={4}
            >
              <Button
                component={'a'}
                variant="contained"
                color="primary"
                size="large"
                fullWidth={isMd ? false : true}
                href={'/crew'}
              >
                Join the flock
              </Button>
              <Box
                marginTop={{ xs: 2, sm: 0 }}
                marginLeft={{ sm: 2 }}
                width={{ xs: '100%', md: 'auto' }}
              >
                <Button
                  component={'a'}
                  href={'/about'}
                  variant="outlined"
                  color="primary"
                  size="large"
                  fullWidth={isMd ? false : true}
                >
                  Learn more
                </Button>
              </Box>
            </Box>
          */}
          </Box>
        </Container>
        <Box
          sx={{
            transform: 'rotate(-20deg)',
            display: { xs: 'none', sm: 'block' },
          }}
        >
          <Box
            display={'flex'}
            width={'50rem'}
            left={'45%'}
            top={0}
            position={'absolute'}
            sx={{ transform: 'translate3d(20%, -40%, 0)' }}
          >
            {images.map((item, i) => (
              <Box key={i} marginTop={{ sm: i * 4 }} marginX={1}>
                {item.group.map((g, j) => (
                  <Box
                    key={j}
                    padding={1}
                    bgcolor={'background.paper'}
                    borderRadius={2}
                    boxShadow={3}
                    marginTop={2}
                  >
                    <Box
                      component={'img'}
                      loading="lazy"
                      src={
                        theme.palette.mode === 'dark'
                          ? g.coverDark ?? g.cover
                          : g.cover
                      }
                      height={1}
                      width={1}
                      maxWidth={320}
                    />
                  </Box>
                ))}
              </Box>
            ))}
          </Box>
        </Box>
      </Box>
      <Box
        component={'svg'}
        preserveAspectRatio="none"
        xmlns="http://www.w3.org/2000/svg"
        x="0px"
        y="0px"
        viewBox="0 0 1920 100.1"
        sx={{
          width: '100%',
          marginBottom: theme.spacing(-1),
        }}
      >
        <path
          fill={theme.palette.background.paper}
          d="M0,0c0,0,934.4,93.4,1920,0v100.1H0L0,0z"
        ></path>
      </Box>
    </Box>
  );
};

export default Hero;
